### PR TITLE
Supporting http headers collection for Gin

### PR DESCRIPTION
**Pull request by IceSoda177**
Supporting http headers collection for Gin

### CODE REVIEW

File: plugins/gin/intercepter.go
Comment (wu-sheng) [inline plugins/gin/intercepter.go:82]: If len==0 then split? Do you mis-write this?

Comment (wu-sheng) [reply plugins/gin/intercepter.go:82]: You should update the gin plugin test to verify this feature.

**Comment by wu-sheng** [comment]
@CodePrometheus Would you like to review this?

File: plugins/gin/intercepter.go
Comment (IceSoda177) [reply plugins/gin/intercepter.go:82]: > If len==0 then split? Do you mis-write this?

They are different config. I use runtimeConfig to store header list, cause the config can't accept array type config param.

Comment (IceSoda177) [reply plugins/gin/intercepter.go:82]: > You should update the gin plugin test to verify this feature.

right

File: tools/go-agent/config/agent.default.yaml
Comment (mrproliu) [inline tools/go-agent/config/agent.default.yaml:101]: I think you missing the configuration declaration. please take a look at other plugins, such as https://github.com/apache/skywalking-go/blob/main/plugins/go-redisv9/config.go

Comment (mrproliu) [reply tools/go-agent/config/agent.default.yaml:101]: And the default value should be `false` I think.

Comment (IceSoda177) [reply tools/go-agent/config/agent.default.yaml:101]: Yes, config.go is missing

**Comment by wu-sheng** [comment]
UT is just a small thing, it is not very useful in agent mode. We have specific plugin test, https://skywalking.apache.org/docs/skywalking-go/next/en/development-and-contribution/write-plugin-testing/

The existing one is at here, https://github.com/apache/skywalking-go/tree/main/test/plugins/scenarios/gin

You should enhance it, verify locally, and update.

**Comment by CodePrometheus** [comment]
> @CodePrometheus Would you like to review this?

I am more than willing to participate in code review, and I will pay attention during my free time at work~ (Beijing time)

**Comment by CodePrometheus** [comment]
@IceSoda177 hello bro, could you help to update the docs/en/agent/plugin-configurations.md? And plz fix the CI.

File: plugins/gin/config.go
Comment (mrproliu) [inline plugins/gin/config.go:22]: A small suggestion, could you change his data type to `[]string`? Although it may be a bit challenging, you might need to provide parsing capabilities in the agent core, so that you don't have to use `runtimeConfig` and reprocess it for each request. If it's still difficult for you, I can help with it. :)

**Comment by IceSoda177** [comment]
> UT is just a small thing, it is not very useful in agent mode. We have specific plugin test, https://skywalking.apache.org/docs/skywalking-go/next/en/development-and-contribution/write-plugin-testing/
> 
> The existing one is at here, https://github.com/apache/skywalking-go/tree/main/test/plugins/scenarios/gin
> 
> You should enhance it, verify locally, and update.

UT is no doubt necessary.

**Comment by IceSoda177** [comment]
> @IceSoda177 hello bro, could you help to update the docs/en/agent/plugin-configurations.md? And plz fix the CI.

Will do

File: plugins/gin/config.go
Comment (IceSoda177) [reply plugins/gin/config.go:22]: Yes, that is bothering me. And I want to solve that. where should I getting start?

**Comment by wu-sheng** [comment]
> > UT is just a small thing, it is not very useful in agent mode. We have specific plugin test, https://skywalking.apache.org/docs/skywalking-go/next/en/development-and-contribution/write-plugin-testing/
> > The existing one is at here, https://github.com/apache/skywalking-go/tree/main/test/plugins/scenarios/gin
> > You should enhance it, verify locally, and update.
> 
> UT is no doubt necessary.

Yes. But it is not verifying the real case. It only verified expected parameter works.

File: plugins/gin/config.go
Comment (mrproliu) [reply plugins/gin/config.go:22]: You need to follow these steps:
1. Adding the `[]string` case in there: https://github.com/apache/skywalking-go/blob/main/tools/go-agent/instrument/plugins/enhance_config.go#L219-L240
2. Adding `ParseStringArray` into the [ToolOperator interface](https://github.com/apache/skywalking-go/blob/main/plugins/core/operator/tools.go) and [implement it](https://github.com/apache/skywalking-go/blob/main/plugins/core/tracer_tools.go).

Comment (IceSoda177) [reply plugins/gin/config.go:22]: Working on it.

**Comment by IceSoda177** [comment]
> UT is just a small thing, it is not very useful in agent mode. We have specific plugin test, https://skywalking.apache.org/docs/skywalking-go/next/en/development-and-contribution/write-plugin-testing/
> 
> The existing one is at here, https://github.com/apache/skywalking-go/tree/main/test/plugins/scenarios/gin
> 
> You should enhance it, verify locally, and update.

The wall is making me crazy...

**Comment by wu-sheng** [comment]
There is nothing we can do about your network environment. And it is common for go dev, we don't add anything new.

**Comment by IceSoda177** [comment]
All modification and test is done, any things else do I need to do?

**Comment by wu-sheng** [comment]
You still miss this,

> The existing one is at here, https://github.com/apache/skywalking-go/tree/main/test/plugins/scenarios/gin
You should enhance it, verify locally, and update.

We need this feature verified in this demo application instrumentation.

**Comment by IceSoda177** [comment]
> all thing, it is not very useful in agent mode. We have specif

Sorry, I get your point now! I will update the scenario later.
My network problem finally solved by using vpn. Otherwise Docker images from google can not be downloaded, and downloading go dependents from github in container will failled.

**Comment by wu-sheng** [comment]
Take your time, it is urgent.

**Review by mrproliu — APPROVED** [review]
LGTM, Once the CI is success, I think It could be merged. :)

**Pull request merged**
